File: src/733/ Flood Fill/Solution.java
class Solution {
    int[] dx = {1, -1, 0, 0};
    int[] dy = {0, 0, 1, -1};
    boolean[][] visited;

    public int[][] floodFill(int[][] image, int sr, int sc, int newColor) {
        int m = image.length, n = image[0].length;
        visited = new boolean[m][n];
        dfs(image, sr, sc, image[sr][sc], newColor);
        return image;
    }

    private void dfs(int[][] image, int sr, int sc, int original, int newColor) {
        image[sr][sc] = newColor;
        visited[sr][sc] = true;

        for (int i = 0; i < 4; i++) {
            int x = sr + dx[i], y = sc + dy[i];
            if (validate(image, x, y, original)) {
                dfs(image, x, y, original, newColor);
            }
        }
    }

    private boolean validate(int[][] image, int x, int y, int original) {
        int m = image.length, n = image[0].length;
        return x >= 0 && x < m && y >= 0 && y < n && image[x][y] == original && !visited[x][y];
    }
}

File: src/733/ Flood Fill/Solution.kt
package `733`.` Flood Fill`

class Solution {
    val direct = arrayOf(intArrayOf(1, 0), intArrayOf(0, 1), intArrayOf(-1, 0), intArrayOf(0, -1))
    fun floodFill(image: Array<IntArray>, sr: Int, sc: Int, newColor: Int): Array<IntArray> {
        if (image[sr][sc] == newColor) return image
        dfs(image, sr, sc, image[sr][sc], newColor)
        return image
    }

    private fun dfs(image: Array<IntArray>, sr: Int, sc: Int, srcColor: Int, newColor: Int) {
        image[sr][sc] = newColor
        (0 until 4).forEach { i ->
            val newX = sr + direct[i][0]
            val newY = sc + direct[i][1]
            if (newX in (0 until image.size) && newY in (0 until image[0].size)
                && image[newX][newY] == srcColor
            ) {
                floodFill(image, newX, newY, newColor)
            }
        }
    }
}